internal/util/vault_sa.go

```diff
@@ -0,0 +1,246 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/libopenstorage/secrets/vault"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	kmsTypeVaultTenantSA = "vaulttenantsa"
+
+	// vaultTenantSAName is the default name of the ServiceAccount that
+	// should be available in the Tenants namespace. This ServiceAccount
+	// will be used to connect to Hashicorp Vault.
+	vaultTenantSAName = "ceph-csi-vault-sa"
+)
+
+/*
+VaultTenantSA represents a Hashicorp Vault KMS configuration that uses a
+ServiceAccount from the Tenant that owns the volume to store/retrieve the
+encryption passphrase of volumes.
+
+Example JSON structure in the KMS config is,
+{
+    "vault-tenant-sa": {
+        "encryptionKMSType": "vaulttenantsa",
+        "vaultAddress": "http://vault.default.svc.cluster.local:8200",
+        "vaultBackendPath": "secret/",
+        "vaultTLSServerName": "vault.default.svc.cluster.local",
+        "vaultCAFromSecret": "vault-ca",
+        "vaultClientCertFromSecret": "vault-client-cert",
+        "vaultClientCertKeyFromSecret": "vault-client-cert-key",
+        "vaultCAVerify": "false",
+        "tenantConfigName": "ceph-csi-kms-config",
+        "tenantSAName": "ceph-csi-vault-sa",
+        "tenants": {
+            "my-app": {
+                "vaultAddress": "https://vault.example.com",
+                "vaultCAVerify": "true"
+            },
+            "an-other-app": {
+                "tenantSAName": "encryped-storage-sa"
+            }
+	},
+	...
+}.
+*/
+type VaultTenantSA struct {
+	vaultTenantConnection
+
+	// tenantSAName is the name of the ServiceAccount in the Tenants Kubernetes Namespace
+	tenantSAName string
+
+	// saTokenDir contains the directory that holds the token to connect to Vault.
+	saTokenDir string
+}
+
+var _ = RegisterKMSProvider(KMSProvider{
+	UniqueID:    kmsTypeVaultTenantSA,
+	Initializer: initVaultTenantSA,
+})
+
+// initVaultTenantSA returns an interface to HashiCorp Vault KMS where Tenants
+// use their ServiceAccount to access the service.
+func initVaultTenantSA(args KMSInitializerArgs) (EncryptionKMS, error) {
+	var err error
+
+	config := args.Config
+	if _, ok := config[kmsProviderKey]; ok {
+		// configuration comes from the ConfigMap, needs to be
+		// converted to vaultTokenConf type
+		config, err = transformConfig(config)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert configuration: %w", err)
+		}
+	}
+
+	kms := &VaultTenantSA{}
+	err = kms.initConnection(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize Vault connection: %w", err)
+	}
+
+	// set default values for optional config options
+	kms.ConfigName = vaultTokensDefaultConfigName
+	kms.tenantSAName = vaultTenantSAName
+
+	// TODO: should this be configurable per tenant?
```

Comment:
I think if the vault namespaces are in place, this looks to be a general configuration, will confirm it though

Reply:
@humblec this was still waiting for a response from you. I have decided that tenants should be able to configure the `vaultRole` option too, it may be needed for tenants to override the default or global configured option.